### PR TITLE
[FLINK-23827][table-planner] Fix ModifiedMonotonicity inference for s…

### PR DESCRIPTION
**Pull request by cshuo**
…ome nodes

## What is the purpose of the change

This pr fixes the problem that some relNodes are not processed properly in `FlinkRelMdModifiedMonotonicity`, which in turn affects the optimizing for Rank.

## Brief change log

*(for example:)*
  - Add/update the handler for some relNodes in `FlinkRelMdModifiedMonotonicity`

## Verifying this change
- Add metadata handler unit test
- Add plan unit test

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by zjuwangg** [comment]
I looked through the code and I think it's tidy and clean way to fix the problem.
Just a little question, how new test in RankTest.scala prove that the fix worked? (Forgive me if this question it's too simple)

**Review by zjuwangg — APPROVED** [review]

**Comment by cshuo** [comment]
Thks @zjuwangg for reviewing. Currently, `ModifiedMonotonicity` is mainly used for inferring the rank strategy, and the added test cannot get `UpdateFastStrategy` for rank before the fix.

Besides, I noticed an unexpected plan for the added test, thks for your reminder.

**Comment by zjuwangg** [comment]
Communicate with @cshuo offline and the new added RankTest cover monotonicity modification. For the UpdateFastRank optimization worked as expected.  
cc @godfreyhe I think this pr can be merged, plz have a another look.

**Comment by zjuwangg** [comment]
@flinkbot run azure

**Comment by zjuwangg** [comment]
I reviewed latest change and thought it's right way to modify test case.

**Comment by zjuwangg** [comment]
@cshuo you may need to rebase the latest code to pass the CI check.

**Comment by zjuwangg** [comment]
cc @godfreyhe have another look and help merge it if everything is ok.

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/RankTest.xml [hunk 1]
```diff
@@ -1020,6 +1062,78 @@ Rank(strategy=[RetractStrategy], rankType=[ROW_NUMBER], rankRange=[rankStart=1,
...
++- Reused(reference_id=[1])
+
+Sink(table=[default_catalog.default_database.sink], fields=[a, b, c])
++- Rank(strategy=[UpdateFastStrategy[0]], rankType=[ROW_NUMBER], rankRange=[rankStart=1, rankEnd=2], partitionBy=[a], orderBy=[b DESC], select=[a, b, c])
+   +- Exchange(distribution=[hash[a]])
+      +- Reused(reference_id=[1])
+]]>
+    </Resource>
+  </TestCase>
+  <TestCase name="testUpdatabletRankWithDeduplicate">
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/RankTest.xml:1103]: nit: typo: Updatablet

**Review by godfreyhe — APPROVED** [review]
LGTM

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Pull request closed without merging**